File: src/cpp/asynchrony/01_basic_async.cpp
/*
This program demonstrates the difference between synchronous and asynchronous
approaches to handling I/O-bound tasks using C++. The synchronous version
runs the tasks sequentially, while the asynchronous version takes advantage of
std::async to run the tasks concurrently. This allows for better resource
utilization and improved performance when dealing with I/O-bound tasks.

The program contains two functions, sync_main() and async_main(), which
demonstrate the synchronous and asynchronous approaches, respectively. The
execution time for each approach is measured and displayed to show the
performance difference.
*/

#include <chrono>
#include <future>
#include <iostream>
#include <thread>

void sync_task(int task_id, int sleep_time) {
  std::cout << "Task " << task_id << " started." << std::endl;
  std::this_thread::sleep_for(std::chrono::seconds(sleep_time));
  std::cout << "Task " << task_id << " finished." << std::endl;
}

void sync_main() {
  auto start_time = std::chrono::steady_clock::now();

  sync_task(1, 2);
  sync_task(2, 1);

  auto elapsed_time = std::chrono::steady_clock::now() - start_time;
  std::cout
      << "Synchronous execution took "
      << std::chrono::duration_cast<std::chrono::seconds>(elapsed_time).count()
      << " seconds." << std::endl;
}

void async_task(int task_id, int sleep_time) {
  std::cout << "Task " << task_id << " started." << std::endl;
  std::this_thread::sleep_for(std::chrono::seconds(sleep_time));
  std::cout << "Task " << task_id << " finished." << std::endl;
}

void async_main() {
  auto start_time = std::chrono::steady_clock::now();

  auto future1 = std::async(std::launch::async, async_task, 1, 2);
  auto future2 = std::async(std::launch::async, async_task, 2, 1);

  future1.wait();
  future2.wait();

  auto elapsed_time = std::chrono::steady_clock::now() - start_time;
  std::cout
      << "Asynchronous execution took "
      << std::chrono::duration_cast<std::chrono::seconds>(elapsed_time).count()
      << " seconds." << std::endl;
}

int main() {
  sync_main();
  async_main();

  return 0;
}

File: src/python/asynchrony/01_basic_async.py
"""
This script demonstrates the difference between synchronous and asynchronous
approaches to handling I/O-bound tasks using Python. The synchronous version
runs the tasks sequentially, while the asynchronous version takes advantage of
asyncio to run the tasks concurrently. This allows for better resource utilization
and improved performance when dealing with I/O-bound tasks.

The script contains two functions, `sync_main()` and `async_main()`, which
demonstrate the synchronous and asynchronous approaches, respectively. The
execution time for each approach is measured and displayed to show the performance
difference.

To see the difference in execution, simply uncomment the appropriate function call
in the `if __name__ == "__main__":` block.
"""

import time
import asyncio


def sync_task(task_id, sleep_time):
    print(f"Task {task_id} started.")
    time.sleep(sleep_time)  # Simulate an I/O-bound task
    print(f"Task {task_id} finished.")


def sync_main():
    start_time = time.time()

    sync_task(1, 2)
    sync_task(2, 1)

    elapsed_time = time.time() - start_time
    print(f"Synchronous execution took {elapsed_time:.2f} seconds.")


async def async_task(task_id, sleep_time):
    print(f"Task {task_id} started.")
    await asyncio.sleep(sleep_time)  # Simulate an I/O-bound task
    print(f"Task {task_id} finished.")


async def async_main():
    start_time = time.time()

    task1 = async_task(1, 2)
    task2 = async_task(2, 1)

    # Run the tasks concurrently
    await asyncio.gather(task1, task2)

    elapsed_time = time.time() - start_time
    print(f"Asynchronous execution took {elapsed_time:.2f} seconds.")


if __name__ == "__main__":
    sync_main()
    asyncio.run(async_main())
